### PR TITLE
Hello dot properties

### PR DESCRIPTION
**Pull request by xavijam**
dot where

The way to introduce dot properties will be adding a new attribute for the simple aggregation, where the user decides if he/she wants to use `marker` or `dot`.

In order to make `dot-*` properties available, those are the necessary tasks:

### v1
- [x] Set new carto reference in all dependencies (`cartodb.js`, `deep-insights.js`, `torque.js`, `tangram-cartocss`, `tangram.cartodb` and `carto` itself) -> https://github.com/CartoDB/carto/pull/34.
- [x] Split big `style-generator` javascript class.
- [x] Add new `method` style component.
- [x] Change way to generate CartoCSS with dot properties. But taking into account:
  - Animated style.
  - Stroke style.
  - Image markers.
  - Autostyle (https://github.com/CartoDB/deep-insights.js/pull/636).
- [x] Add `dot-*` properties in CartoCSS hints.

### v2
- [ ] Decide if dot properties can be used depending on the # of features.

Several parts described above :point_up_2: are already implemented in https://github.com/CartoDB/cartodb/pull/12230/.

- [ ] Check with vector tiles.

cc @rochoa and @matallo.

### CODE REVIEW

**Comment by xavijam** [comment]
This is one possibility for the stroke.

screen shot 2017-11-29 at 17 13 16

I have the CartoCSS generator working but `style-converter.js` is getting bigger, so I'll try to make it clear. Stay tuned!

File: lib/carto/styles/style.rb [hunk 1]
```diff
@@ -21,7 +21,7 @@ def to_cartocss_array
           when 'fill'
             parse_fill(value)
           when 'stroke'
-            parse_stroke(value)
+            parse_stroke(value, @definition['fill'])
```
Comment (xavijam) [inline lib/carto/styles/style.rb]: Not happy with this, just for testing purposes. Find another way.

**Comment by makella** [comment]
@xavijam @rochoa @jorgesancha 

After taking a closer look, I definitely think that dot properties should be reserved for cases when there is a high density of point data, and also based on the amount of data in your current map view.

Unfortunately, for cases where there is less data, the absence of styling properties that you need in those cases, makes a big difference (particularly not having an outline :/ ).

Let's definitely discuss this more! 

I will keep doing some tests, just wanted to point out some initial examples here.

## Examples
### Biggest Cons (for "general purpose" maps)
I definitely think for high density, a lot of this can be overlooked
+ no marker support
+ no transform properties support
  + for example if you, adjusted the height and width of a dot and wanted to rotate it according to an attribute, something like `dot-transform: rotate([field]);` wouldn't be available
+ no outline  

### Outline Attachment 
#### "Outline" Rendering
screen shot 2017-12-08 at 3 43 55 pm

#### Opacity
+ the underlying color shows through and produces a "different" color, so the map does not stay true to its original styling. In the example below, I have the outline attachment color changed to black and applied a `0.5 opacity` to the top layer you can see the result is not a desirable one
  
screen shot 2017-12-08 at 1 29 13 pm

+ One possible work-around here is *basemap dependant* styling. For example, if we had a white attachment outline for Positron and a black attachment outline for Dark Matter, we may be able to make this work. Of course, if a user chooses a different outline color, that won't work :)
screen shot 2017-12-08 at 1 42 20 pm
screen shot 2017-12-08 at 1 40 59 pm

#### Bubble Maps
+ Without a proper outline, the attachment hurts more than it helps where there are overlapping points. For example, in the northeast, I'd want to see a little more differentiation between the overlapping points:
screen shot 2017-12-08 at 1 56 18 pm

+ a possible workaround could be to increase the opacity (a lot), but my worry is if when people filter the data, that the remaining points will be too light:
screen shot 2017-12-08 at 2 05 35 pm

**Comment by xavijam** [comment]
Hey @makella, we are working in another solution, let us finish it and we will show you. We feel it will make happy everybody :). Stay tuned.

**Comment by ramiroaznar** [comment]
@jorgesancha I have asked @matallo to move this to staging to check. Also, @makella should try this as well. 

ccing @csobier @CartoDB/design because this will change the UI if implemented.

**Comment by jorgesancha** [comment]
Thanks @ramiroaznar 

Bear in mind though that these is not definitive. We might use this approach (the two options: better performance vs better cartography, or fast vs slow) to implement the raster aggregation performance improvements the Core Tech guys have been working on.

@rochoa was going to create a couple of tickets so that we can review with Design as we might want to also add some help/tips or guides as to when it makes sense to use each one

**Comment by ramiroaznar** [comment]
Roger that!

**Comment by xavijam** [comment]
Vamos chavales!

**Pull request closed without merging**